builtin/credential/aws/path_login.go

```diff
@@ -888,6 +1049,381 @@ func (b *backend) pathLoginRenew(
+			shortestTTL = maxTTL
+		}
+
+		resp.Auth.TTL = shortestTTL
+	}
+
+	return resp, nil
+}
+
+// These two methods (hasValuesFor*) return two bools
+// The first is a hasAll, that is, does the request have all the values
+// necessary for this auth method
+// The second is a hasAny, that is, does the request have any of the fields
+// exclusive to this auth method
+func hasValuesForEc2Auth(data *framework.FieldData) (bool, bool) {
+	_, hasPkcs7 := data.GetOk("pkcs7")
+	_, hasIdentity := data.GetOk("identity")
+	_, hasSignature := data.GetOk("signature")
+	return (hasPkcs7 || (hasIdentity && hasSignature)), (hasPkcs7 || hasIdentity || hasSignature)
+}
+
+func hasValuesForIamAuth(data *framework.FieldData) (bool, bool) {
+	_, hasRequestMethod := data.GetOk("iam_http_request_method")
+	_, hasRequestUrl := data.GetOk("iam_request_url")
+	_, hasRequestBody := data.GetOk("iam_request_body")
+	_, hasRequestHeaders := data.GetOk("iam_request_headers")
+	return (hasRequestMethod && hasRequestUrl && hasRequestBody && hasRequestHeaders),
+		(hasRequestMethod || hasRequestUrl || hasRequestBody || hasRequestHeaders)
+}
+
+func parseIamArn(iamArn string) (string, string, string, error) {
+	// iamArn should look like one of the following:
+	// 1. arn:aws:iam::<account_id>:user/<UserName>
+	// 2. arn:aws:sts::<account_id>:assumed-role/<RoleName>/<RoleSessionName>
+	// if we get something like 2, then we want to transform that back to what
+	// most people would expect, which is arn:aws:iam::<account_id>:role/<RoleName>
+	fullParts := strings.Split(iamArn, ":")
+	principalFullName := fullParts[5]
+	// principalFullName would now be something like user/<UserName> or assumed-role/<RoleName>/<RoleSessionName>
+	parts := strings.Split(principalFullName, "/")
+	principalName := parts[1]
+	// now, principalName should either be <UserName> or <RoleName>
+	transformedArn := iamArn
+	sessionName := ""
+	if parts[0] == "assumed-role" {
+		transformedArn = fmt.Sprintf("arn:aws:iam::%s:role/%s", fullParts[4], principalName)
+		// fullParts[4] is the <account_id>
+		sessionName = parts[2]
+		// sessionName is <RoleSessionName>
+	} else if parts[0] != "user" {
+		return "", "", "", fmt.Errorf("unrecognized principal type: %q", parts[0])
+	}
+	return transformedArn, principalName, sessionName, nil
+}
+
+func validateVaultHeaderValue(headers http.Header, requestUrl *url.URL, requiredHeaderValue string) error {
+	providedValue := ""
+	for k, v := range headers {
+		if strings.ToLower(iamServerIdHeader) == strings.ToLower(k) {
+			providedValue = strings.Join(v, ",")
+			break
+		}
+	}
+	if providedValue == "" {
+		return fmt.Errorf("didn't find %s", iamServerIdHeader)
+	}
+
+	// NOT doing a constant time compare here since the value is NOT intended to be secret
+	if providedValue != requiredHeaderValue {
```

Comment:
It looks like the `providedValue` will always have a `,` at the end failing to match the `requiredHeaderValue`. No?